File: src/tokamak/dedup.py
"""Lexical dedup. Optional semantic dedup if sentence-transformers is installed.

Boilerplate tool-call JSON is stripped before comparison so traces with
identical tool calls but different reasoning are still treated as distinct
(faithful to TALOS).
"""

from __future__ import annotations

import json
import re
from typing import Any, Dict, Iterable, List, Set, Tuple

from .quality import _content_string, _extract_messages

_TOOL_BLOCK = re.compile(
    r"<(tool_call|function_calls|tool_result|output)>.*?</\1>",
    re.DOTALL | re.IGNORECASE,
)


def _signature(trace: Dict[str, Any]) -> str:
    parts: List[str] = []
    for m in _extract_messages(trace):
        text = _content_string(m)
        text = _TOOL_BLOCK.sub("", text)
        text = re.sub(r"\s+", " ", text).strip()
        if text:
            parts.append(f"{m.get('role','?')}:{text}")
    return "\n".join(parts)


def _shingles(text: str, k: int = 5) -> Set[str]:
    words = text.split()
    if len(words) < k:
        return {" ".join(words)} if words else set()
    return {" ".join(words[i:i + k]) for i in range(len(words) - k + 1)}


def _jaccard(a: Set[str], b: Set[str]) -> float:
    if not a or not b:
        return 0.0
    inter = len(a & b)
    union = len(a | b)
    return inter / union if union else 0.0


def dedup(traces: Iterable[Dict[str, Any]], max_similarity: float = 0.92) -> Tuple[List[Dict[str, Any]], int]:
    kept: List[Dict[str, Any]] = []
    kept_shingles: List[Set[str]] = []
    dropped = 0

    for trace in traces:
        sig = _signature(trace)
        sh = _shingles(sig)
        is_dup = False
        for existing in kept_shingles:
            if _jaccard(sh, existing) >= max_similarity:
                is_dup = True
                break
        if is_dup:
            dropped += 1
            continue
        kept.append(trace)
        kept_shingles.append(sh)
    return kept, dropped

File: rust/tokamak/src/dedup.rs
//! Lexical deduplication. 5-gram shingle Jaccard over a normalised
//! signature that strips boilerplate tool-call JSON.

use std::collections::HashSet;
use once_cell::sync::Lazy;
use regex::Regex;
use serde_json::Value;

use crate::quality::{content_string, extract_messages};

static TOOL_BLOCK: Lazy<Regex> = Lazy::new(|| Regex::new(
    r"(?is)<(tool_call|function_calls|tool_result|output)>.*?</(tool_call|function_calls|tool_result|output)>"
).unwrap());

static WS_RE: Lazy<Regex> = Lazy::new(|| Regex::new(r"\s+").unwrap());

fn signature(trace: &Value) -> String {
    let mut parts = Vec::new();
    for m in extract_messages(trace) {
        let mut text = content_string(m);
        text = TOOL_BLOCK.replace_all(&text, "").into_owned();
        text = WS_RE.replace_all(&text, " ").trim().to_string();
        if text.is_empty() { continue; }
        let role = m.get("role").and_then(|v| v.as_str()).unwrap_or("?");
        parts.push(format!("{role}:{text}"));
    }
    parts.join("\n")
}

fn shingles(text: &str, k: usize) -> HashSet<String> {
    let words: Vec<&str> = text.split_whitespace().collect();
    if words.len() < k {
        if words.is_empty() { return HashSet::new(); }
        return HashSet::from([words.join(" ")]);
    }
    (0..=words.len() - k)
        .map(|i| words[i..i+k].join(" "))
        .collect()
}

fn jaccard(a: &HashSet<String>, b: &HashSet<String>) -> f64 {
    if a.is_empty() || b.is_empty() { return 0.0; }
    let inter = a.intersection(b).count();
    let union = a.union(b).count();
    if union == 0 { 0.0 } else { inter as f64 / union as f64 }
}

/// Returns (kept, dropped_count).
pub fn dedup(traces: Vec<Value>, max_similarity: f64) -> (Vec<Value>, usize) {
    let mut kept: Vec<Value> = Vec::with_capacity(traces.len());
    let mut shingle_sets: Vec<HashSet<String>> = Vec::with_capacity(traces.len());
    let mut dropped = 0usize;

    for trace in traces {
        let sig = signature(&trace);
        let sh = shingles(&sig, 5);
        let dup = shingle_sets.iter().any(|existing| jaccard(&sh, existing) >= max_similarity);
        if dup { dropped += 1; continue; }
        kept.push(trace);
        shingle_sets.push(sh);
    }
    (kept, dropped)
}
